### PR TITLE
chore: reduce circular generation

### PR DESCRIPTION
**Pull request by JoeWang1127**
In this PR:
- Only generate the libraries if the last commit in the current branch has changed `generation_config.yaml`.
- Restore trigger in `.github/workflows/update_googleapis_commit.yaml` and `.github/workflows/hermetic_library_generation.yaml`
- Remove `.github/workflows/generate-from-configuration.yaml`

Example pull request: https://github.com/JoeWang1127/google-cloud-java/pull/29, the `hermetic_library_generation` workflow triggered twice in this PR, but the second time exited early.

### CODE REVIEW

File: generation/hermetic_library_generation.sh [hunk 1]
```diff
@@ -75,6 +75,13 @@ message="chore: generate libraries at $(date)"
 
 git checkout "${target_branch}"
 git checkout "${current_branch}"
+# if the last commit doesn't contain changes to generation configuration,
```
Comment (blakeli0) [inline generation/hermetic_library_generation.sh:78]: In addition to or in place of the comments here, can we add a description to the hermetic_library_generation workflow to describe the behavior?

File: generation/hermetic_library_generation.sh [hunk 2]
```diff
@@ -75,6 +75,13 @@ message="chore: generate libraries at $(date)"
...
+# do not generate again as the result will be the same.
+contains_config_change=$(git diff-tree --no-commit-id --name-only HEAD~1..HEAD -r | grep "${generation_config}")
+if [[ "${contains_config_change}" == "" ]]; then
+    echo "The last commit doesn't contain configuration change, exit."
```
Comment (blakeli0) [inline generation/hermetic_library_generation.sh:82]:
```suggestion
    echo "The last commit doesn't contain any changes to the generation_config.yaml, skipping the whole generation process."
```

File: generation/hermetic_library_generation.sh [hunk 3]
```diff
@@ -102,7 +109,13 @@ docker run \
   --repository-path="${workspace_name}"
 # commit the change to the pull request.
 git add java-* pom.xml gapic-libraries-bom/pom.xml versions.txt "${generation_config}"
-git commit --allow-empty -m "${message}"
+changed_files=$(git diff --cached --name-only)
+if [[ "${changed_files}" == "" ]]; then
+    echo "Nothing to commit, exit."
```
Comment (blakeli0) [inline generation/hermetic_library_generation.sh:114]: Can we provide a more verbose message? Something like "There is no generated code change with the generation config change {insert config diff}."
In addition, this should be a very rare case, right? The only case I can think of is there are googleapis committish change but the change is not for Java.

File: generation/hermetic_library_generation.sh [hunk 4]
```diff
@@ -102,7 +109,13 @@ docker run \
   --repository-path="${workspace_name}"
 # commit the change to the pull request.
 git add java-* pom.xml gapic-libraries-bom/pom.xml versions.txt "${generation_config}"
```
Comment (blakeli0) [inline generation/hermetic_library_generation.sh:111]: I might missed this from last PR, why we need to add `versions.txt` and `generation_config`? I don't think our scripts are touching those files.

File: generation/hermetic_library_generation.sh (hunk 1 above)
Comment (JoeWang1127) [reply generation/hermetic_library_generation.sh:81]: I added a comment at the beginning of this script.

Comment (blakeli0) [reply generation/hermetic_library_generation.sh:81]: Actually I mean the workflow `hermetic_library_generation.yaml` not the script, as that is usually the entry point people start troubleshooting things. But this change looks good too.

File: generation/hermetic_library_generation.sh (hunk 2 above)
Comment (JoeWang1127) [reply generation/hermetic_library_generation.sh]: Done.

File: generation/hermetic_library_generation.sh (hunk 3 above)
Comment (JoeWang1127) [reply generation/hermetic_library_generation.sh]: > Something like "There is no generated code change with the generation config change {insert config diff}."

Done.

> In addition, this should be a very rare case, right? The only case I can think of is there are googleapis committish change but the change is not for Java.

Yes, it is a rare case.

File: generation/hermetic_library_generation.sh (hunk 4 above)
Comment (JoeWang1127) [reply generation/hermetic_library_generation.sh]: I removed `generation_config`.

A new client generation will change `versions.txt` (adding new entry).

File: generation/hermetic_library_generation.sh [hunk 5]
```diff
@@ -75,9 +78,19 @@ message="chore: generate libraries at $(date)"
...
+contains_config_change=$(git diff-tree --no-commit-id --name-only HEAD~1..HEAD -r | grep "${generation_config}")
+if [[ "${contains_config_change}" == "" ]]; then
+    echo "The last commit doesn't contain any changes to the generation_config.yaml, skipping the whole generation process."
+    exit 0
+fi
 # copy generation configuration from target branch to current branch.
 git show "${target_branch}":"${generation_config}" > "${baseline_generation_config}"
-diff "${generation_config}" "${baseline_generation_config}" || echo "config diff"
+config_diff=$(diff "${generation_config}" "${baseline_generation_config}")
+echo "Configuration diff:"
```
Comment (blakeli0) [inline generation/hermetic_library_generation.sh:91]: Can we move this diff logging to right before `git commit` to prevent possible double logging?

**Review by blakeli0 — APPROVED** [review]

**Comment by blakeli0** [comment]
Not sure if it matters, `Verify GAPIC client library generation` is failing, please take a look.

**Comment by JoeWang1127** [comment]
> Not sure if it matters, `Verify GAPIC client library generation` is failing, please take a look.

I don't think it's related to this pull request. It used new-client.py.

File: generation/hermetic_library_generation.sh (hunk 5 above)
Comment (JoeWang1127) [reply generation/hermetic_library_generation.sh]: done.

**Review by diegomarquezp — APPROVED** [review]
IIUC the PR title should be "reduce [circular | cyclical] generation"

**Comment by JoeWang1127** [comment]
We need to update the version in the template.

```
diff --git a/pom.xml b/pom.xml
index 3a[8](https://github.com/googleapis/google-cloud-java/actions/runs/8898222544/job/24434897316?pr=10729#step:5:9)11bd99..aba7f9fe2 100644
--- a/pom.xml
+++ b/pom.xml
@@ -207,7 +207,7 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-deploy-plugin</artifactId>
-        <version>3.1.2</version>
+        <version>3.1.1</version>
         <configuration>
           <skip>true</skip>
         </configuration>
```

**Comment by blakeli0** [comment]
> We need to update the version in the template.
> 
> ```
> diff --git a/pom.xml b/pom.xml
> index 3a[8](https://github.com/googleapis/google-cloud-java/actions/runs/8898222544/job/24434897316?pr=10729#step:5:9)11bd99..aba7f9fe2 100644
> --- a/pom.xml
> +++ b/pom.xml
> @@ -207,7 +207,7 @@
>        <plugin>
>          <groupId>org.apache.maven.plugins</groupId>
>          <artifactId>maven-deploy-plugin</artifactId>
> -        <version>3.1.2</version>
> +        <version>3.1.1</version>
>          <configuration>
>            <skip>true</skip>
>          </configuration>
> ```
> 
> Created [googleapis/sdk-platform-java#2704](https://github.com/googleapis/sdk-platform-java/pull/2704)

I think we should either 
1. Disable the renovate bot config for any Maven dependencies managed in our template.
2. Or extend google-cloud-java's root pom from java-share-config so that we don't have to manage the versions of `maven-deploy-plugin` and `nexus-staging-maven-plugin`.
WDYT @suztomo ?

**Comment by JoeWang1127** [comment]
> I think we should either
> 
> 1. Disable the renovate bot config for any Maven dependencies managed in our template.
> 2. Or extend google-cloud-java's root pom from java-share-config so that we don't have to manage the versions of `maven-deploy-plugin` and `nexus-staging-maven-plugin`.
>    WDYT @suztomo ?

For me, I prefer the 2nd option. 

However, I don't think this issue will block merging this PR as the sdk-platform-java update will fix this issue temporarily and we can implement a fix later.

**Pull request merged**